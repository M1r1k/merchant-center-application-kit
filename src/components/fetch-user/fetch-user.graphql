query LoggedInUserQuery {
  user: me {
    id
    createdAt
    version
    email
    firstName
    gravatarHash
    lastName
    language
    numberFormat
    timeZone
    launchdarklyTrackingId
    launchdarklyTrackingGroup
    launchdarklyTrackingTeam
    defaultProjectKey
    projects {
      total
    }
    organizations {
      total
    }
  }
}
