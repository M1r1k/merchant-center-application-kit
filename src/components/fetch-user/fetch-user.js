import React from 'react';
import PropTypes from 'prop-types';
import { compose, getDisplayName, setDisplayName } from 'recompose';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

const LoggedInUserQuery = gql`
  query LoggedInUser {
    user: me {
      id
      createdAt
      version
      email
      firstName
      lastName
      language
      numberFormat
      timeZone
      tracking_intercom
      intercom_user_hash
      launchdarklyTrackingId
      launchdarklyTrackingGroup
      availableProjects {
        key
        name
        suspended
        expired
      }
      availableOrganizations {
        name
      }
    }
  }
`;

const graphqlOptions = {
  alias: 'withUser',
  name: 'userData',
  options: {
    variables: {
      target: 'mc',
    },
  },
  // Rename `loading` -> `isLoading`, to follow our naming convention
  // https://github.com/commercetools/merchant-center-frontend/issues/2701
  props: ({ userData }) => ({
    userData: {
      isLoading: userData.loading,
      error: userData.error,
      user: userData.user,
    },
  }),
};

// Correct, this is a FaaC component ;)
const FetchUser = props => props.children(props.userData);
FetchUser.displayName = 'FetchUser';
FetchUser.propTypes = {
  children: PropTypes.func.isRequired,
  // Injected
  userData: PropTypes.shape({
    isLoading: PropTypes.bool.isRequired,
    error: PropTypes.object,
    user: PropTypes.object, // see graphql query shape
  }).isRequired,
};

// React component
const FetchLoggedInUser = compose(
  setDisplayName('FetchUser'),
  graphql(LoggedInUserQuery, graphqlOptions)
)(FetchUser);

// HoC
const withUser = mapDataToProps => Component => {
  const WrappedWithUser = props => (
    <FetchLoggedInUser>
      {userData => {
        const mappedProps = mapDataToProps
          ? mapDataToProps(userData)
          : userData;
        return <Component {...props} {...mappedProps} />;
      }}
    </FetchLoggedInUser>
  );
  WrappedWithUser.displayName = `withUser(${getDisplayName(Component)})`;
  return WrappedWithUser;
};

// Public exports
export default FetchLoggedInUser;
export { withUser };

// For testing
export { LoggedInUserQuery, FetchUser };
