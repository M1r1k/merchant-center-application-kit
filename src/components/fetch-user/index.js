export { default, fetchUser } from './fetch-user';
