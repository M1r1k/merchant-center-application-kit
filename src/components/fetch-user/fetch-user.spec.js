import React from 'react';
import { shallow } from 'enzyme';
import LoggedInUserQuery from './fetch-user.graphql';
import { FetchUser, withUser } from './fetch-user';

describe('rendering', () => {
  let props;
  let wrapper;

  describe('<FetchUser>', () => {
    beforeEach(() => {
      props = {
        children: jest.fn(),
        // this is usually injected by graphql
        userData: {
          isLoading: false,
          user: {
            firstName: 'John',
          },
        },
      };
      wrapper = shallow(<FetchUser {...props} />);
    });
    it('should call children with userData object', () => {
      expect(props.children).toHaveBeenCalledWith(props.userData);
    });
  });

  describe('withUser()', () => {
    let wrapperRender;

    describe('when mapDataToProps is defined', () => {
      beforeEach(() => {
        // eslint-disable-next-line react/prop-types
        const Profile = ({ firstName }) => <div>{firstName}</div>;
        const ProfileWithUser = withUser(userData => ({
          firstName: userData.user.firstName,
        }))(Profile);
        wrapper = shallow(<ProfileWithUser foo="bar" />);
        wrapperRender = shallow(
          wrapper.prop('children')({ user: { firstName: 'John' } })
        );
      });
      it('should render <Profile> with firstName', () => {
        expect(wrapperRender).toMatchElement(<div>{'John'}</div>);
      });
      it('should render FetchUser internally', () => {
        expect(wrapper).toRender('FetchUser');
      });
    });
    describe('when mapDataToProps is not defined', () => {
      beforeEach(() => {
        // eslint-disable-next-line react/prop-types
        const Profile = ({ user }) => <div>{user.firstName}</div>;
        const ProfileWithUser = withUser()(Profile);
        wrapper = shallow(<ProfileWithUser foo="bar" />);
        wrapperRender = shallow(
          wrapper.prop('children')({ user: { firstName: 'John' } })
        );
      });
      it('should render <Profile> with firstName', () => {
        expect(wrapperRender).toMatchElement(<div>{'John'}</div>);
      });
      it('should render FetchUser internally', () => {
        expect(wrapper).toRender('FetchUser');
      });
    });
  });

  describe('graphql query', () => {
    it('should match snapshot', () => {
      expect(LoggedInUserQuery).toMatchSnapshot();
    });
  });
});
