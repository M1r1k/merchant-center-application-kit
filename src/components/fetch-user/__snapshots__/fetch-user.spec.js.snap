// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering graphql query should match snapshot 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "LoggedInUser",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": Object {
              "kind": "Name",
              "value": "user",
            },
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "me",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "createdAt",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "version",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "email",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "firstName",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "lastName",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "language",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "numberFormat",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "timeZone",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "tracking_intercom",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "intercom_user_hash",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "launchdarklyTrackingId",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "launchdarklyTrackingGroup",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "availableProjects",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "key",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "suspended",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "expired",
                        },
                        "selectionSet": null,
                      },
                    ],
                  },
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "availableOrganizations",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": null,
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 435,
    "start": 0,
  },
}
`;
